modules/apps/message-boards/message-boards-service/src/main/java/com/liferay/message/boards/internal/util/MBMessageNotificationTemplateHelper.java

```diff
@@ -0,0 +1,351 @@
+ */
+
+package com.liferay.message.boards.internal.util;
+
+import com.liferay.message.boards.constants.MBMessageConstants;
+import com.liferay.message.boards.model.MBMessage;
+import com.liferay.message.boards.service.MBMessageLocalService;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.exception.PortalException;
+import com.liferay.portal.kernel.feature.flag.FeatureFlagManagerUtil;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.parsers.bbcode.BBCodeTranslatorUtil;
+import com.liferay.portal.kernel.service.ServiceContext;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.HashMapBuilder;
+import com.liferay.portal.kernel.util.ListUtil;
+import com.liferay.portal.kernel.util.Validator;
+import com.liferay.portal.kernel.util.WebKeys;
+import com.liferay.portal.kernel.workflow.WorkflowConstants;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.servlet.http.HttpServletRequest;
+
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * @author Alicia García
+ */
+public class MBMessageNotificationTemplateHelper {
+
+	public MBMessageNotificationTemplateHelper(
+		boolean htmlFormat, int maxNumberOfMessages,
+		int maxNumberOfThreadMessages,
+		MBMessageLocalService mbMessageLocalService,
+		ServiceContext serviceContext) {
+
+		_htmlFormat = htmlFormat;
+		_maxNumberOfMessages = maxNumberOfMessages;
+		_maxNumberOfThreadMessages = maxNumberOfThreadMessages;
+		_mbMessageLocalService = mbMessageLocalService;
+		_serviceContext = serviceContext;
+	}
+
+	public String getMessageBody(MBMessage mbMessage, String quoteMark) {
+		if (!_htmlFormat) {
+			return _getQuotedMessage(true, mbMessage.getBody(), quoteMark);
+		}
+
+		if (!mbMessage.isFormatBBCode()) {
+			return mbMessage.getBody();
+		}
+
+		try {
+			String messageBody = BBCodeTranslatorUtil.getHTML(
+				mbMessage.getBody());
+
+			HttpServletRequest httpServletRequest =
+				_serviceContext.getRequest();
+
+			if (httpServletRequest == null) {
+				return messageBody;
+			}
+
+			ThemeDisplay themeDisplay =
+				(ThemeDisplay)httpServletRequest.getAttribute(
+					WebKeys.THEME_DISPLAY);
+
+			return MBUtil.replaceMessageBodyPaths(themeDisplay, messageBody);
+		}
+		catch (Exception exception) {
+			_log.error(
+				StringBundler.concat(
+					"Unable to parse mbMessage ", mbMessage.getMessageId(),
+					": ", exception.getMessage()));
+		}
+
+		return mbMessage.getBody();
+	}
+
+	public String renderMessageParent(MBMessage parentMessage) {
+		if (!FeatureFlagManagerUtil.isEnabled("LPS-182020") ||
+			(_maxNumberOfThreadMessages == 0) ||
+			(parentMessage.getParentMessageId() ==
+				MBMessageConstants.DEFAULT_PARENT_MESSAGE_ID)) {
+
+			return StringPool.BLANK;
+		}
+
+		List<MBMessage> mbMessages = new ArrayList<>();
```

Comment:
As you're always adding at the front of the list, it is better to use a `LinkedList`. Internally, `ArrayList` uses an array, so adding at the front requires to shift all remaining elements.